### PR TITLE
Turns squashing mobs into an element

### PR DESCRIPTION
**Pull request by ghost**
:cl: Qustinnus
 code: squashing behavior is now an element, now anything could theoretically be squished!
 /:cl:

Opens up room for extending squashing behavior for future cytology updates.

allows for shit like https://streamable.com/8n6ybn

### CODE REVIEW

**Comment by cacogen** [comment]
Damn that example is fucking cool. Reminds me of Streets of Rogue. Seems too fun for /tg/station coders to allow though

File: code/datums/elements/squashable.dm [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	if(!src.on_squash_callback && squash_callback)
+		on_squash_callback = CALLBACK(target, squash_callback)
+
+	RegisterSignal(target, COMSIG_MOVABLE_CROSSED, .proc/OnCrossed)
+
+/datum/element/squashable/Detach(mob/living/target)
+	UnregisterSignal(target, COMSIG_MOVABLE_CROSSED)
+
+///Handles the squashing of the mob
+/datum/element/squashable/proc/OnCrossed(mob/living/target, atom/movable/AM)
```
Comment (Jared-Fogle) [inline code/datums/elements/squashable.dm:33]: Better variable names.

File: code/datums/elements/squashable.dm [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+			if(HAS_TRAIT(crossing_mob, TRAIT_PACIFISM))
+				crossing_mob.visible_message("<span class='notice'>[crossing_mob] carefully steps over [target].</span>", "<span class='notice'>You carefully step over [target] to avoid hurting it.</span>")
+				return
+			if(should_squash)
+				crossing_mob.visible_message("<span class='notice'>[crossing_mob] squashed [target].</span>", "<span class='notice'>You squashed [target].</span>")
+				Squish(target)
+			else
+				target.visible_message("<span class='notice'>[target] avoids getting crushed.</span>")
+	else if(isstructure(AM))
+
```
Comment (Jared-Fogle) [inline code/datums/elements/squashable.dm:57]:
```suggestion
```

File: code/modules/mob/living/simple_animal/hostile/cockroach.dm [hunk 3]
```diff
@@ -36,6 +36,10 @@
 /mob/living/simple_animal/hostile/cockroach/Initialize()
 	. = ..()
 	add_cell_sample()
+	make_squashable()
+
+/mob/living/simple_animal/hostile/cockroach/proc/make_squashable()
+	AddElement(/datum/element/squashable, 50, 1)
```
Comment (Jared-Fogle) [inline code/modules/mob/living/simple_animal/hostile/cockroach.dm:42]: Named arguments, if possible.

File: code/modules/mob/living/simple_animal/hostile/cockroach.dm [hunk 4]
```diff
@@ -110,16 +94,21 @@
...
-		else if(HAS_TRAIT(A, TRAIT_PACIFISM))
-			A.visible_message("<span class='notice'>[A] carefully steps over [src].</span>", "<span class='notice'>You carefully step over [src] to avoid hurting it.</span>")
-		else if(HAS_TRAIT(A, TRAIT_PIERCEIMMUNE))
-			A.visible_message("<span class='danger'>[A] steps onto [src]'s spike, but is unfazed!</span>", "<span class='userdanger'>You step onto [src]'s spike, but you're unaffected!</span>")
-		else
-			A.visible_message("<span class='notice'>[A] squashes [src], not even noticing its spike.</span>", "<span class='notice'>You squashed [src], not even noticing its spike.</span>")
-			adjustBruteLoss(1) //kills a normal cockroach
-	..()
+/mob/living/simple_animal/hostile/cockroach/hauberoach/make_squashable()
+	AddElement(/datum/element/squashable, 100, 1, null, /mob/living/simple_animal/hostile/cockroach/hauberoach/.proc/on_squish)
```
Comment (Jared-Fogle) [inline code/modules/mob/living/simple_animal/hostile/cockroach.dm:98]: Named arguments here too, if possible.

File: code/modules/mob/living/simple_animal/hostile/cockroach.dm [hunk 5]
```diff
@@ -110,16 +94,21 @@
...
+
+///Proc used to override the squashing behavior of the normal cockroach.
+/mob/living/simple_animal/hostile/cockroach/hauberoach/proc/on_squish(mob/living/cockroach, atom/movable/AM)
```
Comment (Jared-Fogle) [inline code/modules/mob/living/simple_animal/hostile/cockroach.dm:101]: Better variable name (even just `_living_target`).

File: code/datums/elements/squashable.dm [hunk 6]
```diff
@@ -0,0 +1,68 @@
+/datum/element/squashable
```
Comment (Jared-Fogle) [inline code/datums/elements/squashable.dm:1]: Document this.

**Review by Mothblocks — CHANGES_REQUESTED** [review]

**Comment by Qustinnus** [comment]
done

File: code/modules/mob/living/simple_animal/hostile/cockroach.dm [hunk 7]
```diff
@@ -110,16 +94,20 @@
...
-		else if(HAS_TRAIT(A, TRAIT_PACIFISM))
-			A.visible_message("<span class='notice'>[A] carefully steps over [src].</span>", "<span class='notice'>You carefully step over [src] to avoid hurting it.</span>")
-		else if(HAS_TRAIT(A, TRAIT_PIERCEIMMUNE))
-			A.visible_message("<span class='danger'>[A] steps onto [src]'s spike, but is unfazed!</span>", "<span class='userdanger'>You step onto [src]'s spike, but you're unaffected!</span>")
-		else
-			A.visible_message("<span class='notice'>[A] squashes [src], not even noticing its spike.</span>", "<span class='notice'>You squashed [src], not even noticing its spike.</span>")
-			adjustBruteLoss(1) //kills a normal cockroach
-	..()
+/mob/living/simple_animal/hostile/cockroach/hauberoach/make_squashable()
+	AddElement(/datum/element/squashable, squash_chance = 100, squahs_damage = 1, squash_callback = /mob/living/simple_animal/hostile/cockroach/hauberoach/.proc/on_squish)
```
Comment (Rohesie) [inline code/modules/mob/living/simple_animal/hostile/cockroach.dm:98]: squahs

Comment (Jared-Fogle) [reply code/modules/mob/living/simple_animal/hostile/cockroach.dm:98]: prime lint target 😏

**Comment by Rohesie** [comment]
Test it locally. If it's working as intended this looks ready for merge.

**Comment by Qustinnus** [comment]
done

**Pull request merged**